rust/cymbal/src/pipeline/mod.rs

```diff
@@ -0,0 +1,176 @@
+use std::sync::Arc;
+
+use billing::apply_billing_limits;
+use chrono::{DateTime, NaiveDateTime, Utc};
+use clean::clean_set_props;
+use common_kafka::{
+    kafka_messages::ingest_warning::IngestionWarning, kafka_producer::send_iter_to_kafka,
+};
+use common_types::{CapturedEvent, ClickHouseEvent};
+
+use exception::do_exception_handling;
+use geoip::add_geoip;
+use person::add_person_properties;
+use prep::prepare_events;
+use serde::Deserialize;
+
+use crate::{
+    app_context::AppContext,
+    error::{EventError, PipelineFailure, PipelineResult, UnhandledError},
+    metric_consts::{
+        BILLING_LIMITS_TIME, CLEAN_PROPS_TIME, EMIT_INGESTION_WARNINGS_TIME,
+        EXCEPTION_PROCESSING_TIME, GEOIP_TIME, PERSON_PROCESSING_TIME, PREPARE_EVENTS_TIME,
+        TEAM_LOOKUP_TIME,
+    },
+    teams::do_team_lookups,
+};
+
+pub mod billing;
+pub mod clean;
+pub mod errors;
+pub mod exception;
+pub mod geoip;
+pub mod person;
+pub mod prep;
+
+// We can receive either ClickhouseEvents or CaptureEvents
+#[derive(Debug, Clone, Deserialize)]
+// ClickhouseEvent is hefty compared to CapturedEvent (496 vs 160 bytes), but we mostly pass
+// around vecs of these and I'd rather skip the pointer chase
+#[allow(clippy::large_enum_variant)]
+pub enum IncomingEvent {
+    ClickhouseReady(ClickHouseEvent),
+    Captured(CapturedEvent),
+}
+
+pub async fn handle_batch(
+    buffer: Vec<IncomingEvent>,
+    context: Arc<AppContext>,
+) -> Result<Vec<PipelineResult>, PipelineFailure> {
```

Comment:
Having the entire pipeline inside cymbal feels somehow weird. Is there any plans to abstract it out to something shared?

I'm basically worried about the case where more people want to have a similar ingestion service. Then we have many different ingestion paths each with variations. I'd worry that they don't all get updated and we're forever chasing subtle differences. At least the plugin server has a single runner file